### DISCUSSION TITLE
When does preallocation=1 (fast) allocate, and how much?

### DISCUSSION THREAD

**Discussion by zwimer** [transmission/transmission#7151] [Q&A] [answered]
I understand what no preallocation and full preallocation do, but what does preallocation=1 ('fast') do? As in, what does it preallocate and when? I'm looking at this to better understand how fragmented a download might be when `preallocation=1` is set.

For example: Does it preallocate a 1GB chunk every time a piece needs a new chunk? Does it pre-allocate per piece? Does it do a full allocation after data spooled RAM is too large?

**Comment by reardonia** [comment]
> I understand what no preallocation and full preallocation do, but what does preallocation=1 ('fast') do? As in, what does it preallocate and when? I'm looking at this to better understand how fragmented a download might be when `preallocation=1` is set.


What it does is dependent on the underlying filesystem and OS.  some os+fs combos support fast preallocation (blocks are allocated, typically as contiguously as possible, without actually writing the blocks out) and some don't.  On typical Linux installations, Transmission will make the call to preallocate and rely on the OS to do the right thing. But that could also result in a very long preallocation.

**Reply by zwimer** [reply]
I do understand that what happens when transmission attempts to pre-allocate space is up to the OS and FS, such as COW file systems just setting up internal pointers to zero'd extents rather than actually allocating data- my question though is what the difference is between `preallocation=2` (full) allocation and `preallocation=1` (fast). For example, assuming a Kernel/FS that will preallocate if requested, I imagine a full preallocation would immediately pre-allocate the entire file.

For example: Is the difference that full allocation forces a zero-initialization of the data where as fast allocate might allow marking the blocks as uninitialized (such as how `fallocate` attempts to)? Or something else?

**Comment by reardonia** [comment, marked as answer]
Sorry if I didn't presume you knew about the saga of preallocation on Posix ;)

At least on Unix/Linux, setting to 'full' will call `posix_fallocate()`.  Setting to 'sparse' or 'fast' will call `fallocate64()`.  'full' does not guarantee that zeroes will be written, it depends on the underlying Posix implementation.  The code in file-posix.cc and open-files.cc has notes explaining that this hasn't been verified in quite some time.

Summary: you may get the same behavior for 'full' and 'fast' (which is synonymous with 'sparse') or you may get different behavior depending on the implementation of `posix_fallocate`.  Likely that it'll be the same on a modern system. Should note that there is a Mac-specific preallocator as well, which can stand in for `posix_fallocate` on that platform.

(I'll play with this on a few systems tomorrow to see what happens)

**Comment by JustMyGithub** [comment]
ext4 can do sparse, like "truncate -s 5G testfile.bin" shows. Why does transmission follow some ancient unix/posix definition instead of doing what the users expect it to do (here containtuitively do full allocation if preallocate is set to 1) ?

OK, its even worse. even with preallocate = 0 ist does a full allocation. so the preallocation setting is absolutely useless?!

**Reply by reardonia** [reply]
> ext4 can do sparse, like "truncate -s 5G testfile.bin" shows. Why does transmission follow some ancient unix/posix definition instead of doing what the users expect it to do (here containtuitively do full allocation if preallocate is set to 1) ?
> 
> OK, its even worse. even with preallocate = 0 ist does a full allocation. so the preallocation setting is absolutely useless?!

You misunderstand how this works.  Set it to =1, you'll be fine.

**Reply by JustMyGithub** [reply]
No. I tested 1 (default), and even 0. it always does full alloc.

**Comment by hialexk** [comment]
I don’t understand the underlying code principles, but in practical use, I’ve found that Transmission’s preallocation feature is extremely awkward. It has significant flaws.
我不懂代码原理，但在实际使用中发现，tr的预分配功能非常难用，存在巨大缺陷。

【1】In practical use under mode 1, I noticed a frustrating behavior: when I selectively choose to download only a few specific files from a large torrent, Transmission often still creates preallocated placeholder files and reserves full disk space for the unselected files. If the drive is low on space, I have to manually go into the directory and delete these unwanted empty files, which is extremely tedious—especially when dealing with massive multi-file torrents. 
在模式 1 的实际使用中，我注意到一个令人沮丧的问题：当我从大型种子文件中选择性地只下载几个特定文件时，Transmission 仍然经常会创建预分配的占位符文件，并为未选择的文件预留完整的磁盘空间。如果磁盘空间不足，我不得不手动进入目录并删除这些不需要的空文件，这非常繁琐——尤其是在处理包含大量文件的大型种子文件时。

【2】As a Windows user, setting preallocation to 2 is highly impractical because Transmission physically writes zeros to the disk, causing extreme disk wear and locking the active time at 100%.
作为 Windows 用户，将 preallocation 设置为 2 非常不切实际，因为 Transmission 会将零物理写入磁盘，导致磁盘严重磨损，并将活动时间锁定在 100%。

Other popular Windows clients such as qBittorrent and BitComet fully support fast preallocation on NTFS（Similar to Mode 2
in tr）without physically zero-filling the drive. 
其他一些流行的 Windows 客户端，例如 qBittorrent 和 BitComet，都完全支持 NTFS 文件系统上的快速预分配（类似tr的模式2），而无需将磁盘物理填充为零。

【3】When Transmission uses Mode 1 fast preallocation for large torrents, the client freezes with 100% disk I/O for over ten minutes if the logical preallocated size exceeds drive free space.
当 Transmission 使用模式 1 快速预分配来处理大型种子时，如果逻辑预分配的大小超过驱动器可用空间，客户端将冻结超过 10 分钟，磁盘 I/O 使用率达到 100%。

-----------
https://github.com/transmission/transmission/issues/8902
I created this issue but haven’t received any response. In recent years, the development team has mainly focused on fixing bugs and refactoring the codebase stemming from Transmission’s transition from 3.0 to 4.0. They rarely add new features or make substantial changes to low-level functionality.
我提出了这个ISSUE，但没有收到回复。近些年开发组主要处理tr从3.0转到4.0后产生的bug、代码设计，很少增加新功能、或者大幅改动底层功能。
